Spatial.gd

```diff
@@ -420,80 +425,110 @@ func gen_map_markers():
-		category_item.set_metadata(0,category)
-		category_item.set_cell_mode(1, TreeItem.CELL_MODE_CHECK)
-		category_item.set_checked(1, self.local_data.visible_dict.get(tree_path, false))
-		category_item.set_tooltip(1, "Show/Hide")
-		category_item.set_editable(1, true)
-		category_item.set_collapsed(collapsed)
-		for category_child in category.get_children():
-			add_category(category_item, category_child, tree_path + "." + category_child.get_name(), true)
+func add_category(item: TreeItem, category, full_category_name: String, collapsed: bool):
+	if category.get_name() == "": 
+		print("Category found with no name. Full name ", full_category_name)
+		return
+	var category_item = self.marker_packs.create_item(item)
+	category_item.set_text(0,category.get_display_name())
+	category_item.set_metadata(0, full_category_name)
+	category_item.set_cell_mode(1, TreeItem.CELL_MODE_CHECK)
+	category_item.set_checked(1, Settings.local_category_data.get(full_category_name, {}).get("checked", false))
+	category_item.set_tooltip(1, "Show/Hide")
+	category_item.set_editable(1, true)
+	category_item.set_collapsed(collapsed)
+	for category_child in category.get_children():
+		add_category(category_item, category_child, full_category_name + "." + category_child.get_name(), true)
 
 
 func switch_selected_category(category_item: TreeItem):
-	var selected_category_name = find_pedigree_name(category_item, category_item.get_metadata(0).get_name())
-	local_data.visible_dict[selected_category_name] = category_item.is_checked(1)
-	switch_checkboxes(category_item, category_item.is_checked(1))
-	
-	node_checkmark_switch(selected_category_name, category_item, self.paths)
-	node_checkmark_switch(selected_category_name, category_item, self.minimap)
-	node_checkmark_switch(selected_category_name, category_item, self.icons)
-
-	$Control/LocalData.save_local()
+	Settings.local_category_data[category_item.get_metadata(0)] = {
+		"checked" : category_item.is_checked(1),
+	} 
+	update_node_visibility(category_item, self.paths)
+	update_node_visibility(category_item, self.icons)
 
+	Settings.save()
 
-func find_pedigree_name(category_item: TreeItem, pedigree_name: String):
-	var parent =  category_item.get_parent()
-	if parent.is_selectable(0):
-		pedigree_name = parent.get_metadata(0).get_name() + "." + pedigree_name
-		pedigree_name = find_pedigree_name(parent, pedigree_name)
-	return pedigree_name
-
-
-func node_checkmark_switch(selected_category_name: String, category_item, nodes):
+func update_node_visibility(category_item: TreeItem, nodes):
 	for node in nodes.get_children():
-		if node.category_name.begins_with(selected_category_name):
-			local_data.visible_dict[node.category_name] = category_item.is_checked(1)
-			if category_item.is_checked(1):
-				node.show()
+		var node_name = node.waypoint.get_category().get_name()
+		if node_name.begins_with(category_item.get_metadata(0)):
+			if is_category_visible(category_item):
```

Comment:
It should be the nodes category. Reworked to avoid this. I now make a dictionary with the keys being the affected categories and the value being the results of `is_category_visible()` and compare the nodes to this dictionary.